src/main/java/net/starlark/java/eval/StarlarkInt.java

```diff
@@ -512,10 +514,59 @@ public static StarlarkInt multiply(StarlarkInt x, StarlarkInt y) {
       return StarlarkInt.of(xl * yl);
     }
 
+    try {
+      long xl = x.toLongFast();
+      long yl = y.toLongFast();
+
+      // see Hacker's Delight, 8-2
+
+      long u0 = xl & 0xFFFFFFFFL;
+      long u1 = xl >> 32;
+      long v0 = yl & 0xFFFFFFFFL;
+      long v1 = yl >> 32;
+      long w0 = u0 * v0;
+      long t = u1 * v0 + (w0 >>> 32);
+      long w1 = t & 0xFFFFFFFFL;
+      long w2 = t >> 32;
+      w1 = u0 * v1 + w1;
+
+      // high and low 64-bits of 128-bit signed multiplication result
+      long high = u1 * v1 + w2 + (w1 >> 32);
+      long low = xl * yl;
+
+      // checking 128-bit signed integer is within 64-bit signed integer range
+      if ((high == 0 && (low & Long.MIN_VALUE) == 0) || (high == -1 && (low & Long.MIN_VALUE) != 0)) {
+        long zl = low | (high & Long.MIN_VALUE);
+        return StarlarkInt.of(zl);
+      }
+
+      /* overflow */
+
+    } catch (Overflow unused) {
+      /* fall through */
+    }
+
+    // avoid unnecessary conversion to BigInteger if the other argument is -1, 0, 1
+    // also make self-test below faster
+    if (x == ZERO || y == ONE) {
```

Comment:
Add a comment here explaining that this is merely a fast-path optimization.
of() does not guarantee to return canonical objects (there's a potential race in
populating the memoization table).

Reply:
> of() does not guarantee to return canonical objects

It guarantees when `of` is called from `static final`.

> this is merely a fast-path optimization

It is not, assertion below relies on it.

Anyway, I'll rewrite it using `equals` and no globals.

Reply:
I realized later that we already rely on identity for ZERO in other places, so I retract my concern. As you point out, if all the calls are from the static initializer of a single class, we're fine.

Feel free to restore the == tests.